### PR TITLE
Sync: Create new time entry state (POST)

### PR DESCRIPTION
**Pull request by simonrozsival**
Ref #498 

This PR contains a state for creating a new time entry (HTTP POST) and then persists the data received from the server to the local database.

### CODE REVIEW

File: Toggl.Foundation/Sync/States/Push/BaseCreateEntityState.cs [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+
+namespace Toggl.Foundation.Sync.States
+{
+    internal abstract class BaseCreateEntityState<TModel>
+        where TModel : class, IBaseModel, IDatabaseSyncable
+    {
+        private readonly ITogglApi api;
+        private readonly ITogglDatabase database;
+
+        public StateResult<TModel> CreatingFailed { get; } = new StateResult<TModel>();
```
Comment (amulware) [inline Toggl.Foundation/Sync/States/Push/BaseCreateEntityState.cs]: I'm fine doing this later, but this has to pass on the exception as well, especially for API exceptions.

File: Toggl.Foundation.Tests/Sync/States/Push/BaseCreateEntityStateTests.cs [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+        public void ReturnsFailTransitionWhenHttpFails()
+            => helper.ReturnsFailTransitionWhenHttpFails();
+
+        [Fact]
+        public void ReturnsFailTransitionWhenDatabaseOperationFails()
+            => helper.ReturnsFailTransitionWhenDatabaseOperationFails();
+
+        [Fact]
+        public void ReturnsSuccessfulTransitionWhenEveyrthingWorks()
+            => helper.ReturnsSuccessfulTransitionWhenEveyrthingWorks();
```
Comment (amulware) [inline Toggl.Foundation.Tests/Sync/States/Push/BaseCreateEntityStateTests.cs]: Typos

File: Toggl.Foundation/Sync/States/Push/BaseCreateEntityState.cs [hunk 3]
```diff
@@ -12,7 +12,7 @@ internal abstract class BaseCreateEntityState<TModel>
         private readonly ITogglApi api;
         private readonly ITogglDatabase database;
 
-        public StateResult<TModel> CreatingFailed { get; } = new StateResult<TModel>();
+        public StateResult<PushFailure<TModel>> CreatingFailed { get; } = new StateResult<PushFailure<TModel>>();
```
Comment (amulware) [inline Toggl.Foundation/Sync/States/Push/BaseCreateEntityState.cs]: Why not use a tuple?

**Comment by amulware** [comment]
Comment shows as outdated because I made it on a commit, even though it is still relevant. Github, you silly.

File: Toggl.Foundation/Sync/States/Push/BaseCreateEntityState.cs (hunk 3 above)
Comment (simonrozsival) [reply Toggl.Foundation/Sync/States/Push/BaseCreateEntityState.cs]: I first did this with tuples and then decided to make it a separate class to make it more convenient for sharing between the states. I like this second approach more than using tuples when the states are so loosely coupled.

**Pull request merged**